Feature/15429 domains dashboard logic improvements

ravishanker opened the pull request:
Fixes #15429 

This PR does the following 

- Shows Primary domain at the top and below it one of the following
- Shows **Your site domains** List if the site already has custom domain/s
- Shows **Get your domain** if it is a free plan and/or no domain credit.  
- Shows **Claim your free domain** if site has domain credit

This doesn't match with what is listed here entirely p1633549036242800-slack-C0280085R7W.   I'm not sure what to follow here.  Expect to fix it if any in the next PR after clarification.

<table><tr><td><img width=320 src="https://user-images.githubusercontent.com/990349/136909187-a520973c-a59b-4f1d-a169-2dff4fda42e1.png"</td><td><img width=320 src="https://user-images.githubusercontent.com/990349/136332736-0e6d4170-edb0-4496-b43d-24cda05a9dfa.png"</td><td><img width=320 src="https://user-images.githubusercontent.com/990349/136332759-c27c775b-eb17-40e2-b59a-37869ef8def2.png"</td></tr></table>

Fixes #15234 

To test:

- Enable AppSettings -> Debug settings -> SiteDomainsFeatureConfig ✅
- Return to My Site, Notice there's a new item under Configuration -> 🌐 Domains
- Click on Domains, takes you to **Domains** dashboard screen
- Test with a site that has free plan, and a paid plan

To test:

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing and unit tests

3. What automated tests I added (or what prevented me from doing so)
Unit tests

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ravishanker committed:
Update dashboard logic

Refactored dashboard logic to handle custom domains and domain credit conditions

ravishanker committed:
Merge branch 'develop' into feature/15429-domains-dashboard-logic-improvements

ravishanker committed:
Merge branch 'develop' into feature/15429-domains-dashboard-logic-improvements

ravishanker committed:
Update DomainsDashboardViewModel.kt

remove logs

ravishanker committed:
Create DomainsDashboardViewModelTest.kt

added unit test for view model

ravishanker committed:
fix lint issues

fixed spacing

thehenrybyrd commented:
Hi @ravishanker I'm available to test this PR today/tomorrow, but wanted to make sure I understand the limitations before doing so.

**Here's what I would test:**
Check Domains Dashboard on multiple sites with different setups.
- Verify that the “Your site domains” list shows if the site has any custom domains.
- Verify that “Get your domain” appears for sites with no domain credit.
- Verify that “Claim your free domain” appears for sites with a domain credit.

**These items would be Out of Scope:** (known issues and incomplete work) 
- Domains list may not be accurate; this PR doesn’t guarantee valid data is displayed.
- Some scenarios detailed by Renan need to be handled still; _this test would specify which_.
- Domains Dashboard UI not fully implemented.

**I would plan to check the 8 scenarios Renan described:**
Free plan with custom domain 
Free plan with no custom domain 
Paid plan with primary custom domain and credits
Paid plan with primary custom domain and no credits
Paid plan with secondary custom domain and credits 
Paid plan with secondary custom domain and no credits 
Paid plan with no custom domain and credits 
Paid plan with no custom domain and no credits  

Let me know if that would be helpful to you, and if I'm missing anything about the scope of the PR.

ravishanker commented:
> Hi @ravishanker I'm available to test this PR today/tomorrow, but wanted to make sure I understand the limitations before doing so.
> 
> **Here's what I would test:** Check Domains Dashboard on multiple sites with different setups.
> 
>     * Verify that the “Your site domains” list shows if the site has any custom domains.
> 
>     * Verify that “Get your domain” appears for sites with no domain credit.
> 
>     * Verify that “Claim your free domain” appears for sites with a domain credit.
> 
> 
> **These items would be Out of Scope:** (known issues and incomplete work)
> 
>     * Domains list may not be accurate; this PR doesn’t guarantee valid data is displayed.
> 
>     * Some scenarios detailed by Renan need to be handled still; _this test would specify which_.
> 
>     * Domains Dashboard UI not fully implemented.
> 
> 
> **I would plan to check the 8 scenarios Renan described:** Free plan with custom domain Free plan with no custom domain Paid plan with primary custom domain and credits Paid plan with primary custom domain and no credits Paid plan with secondary custom domain and credits Paid plan with secondary custom domain and no credits Paid plan with no custom domain and credits Paid plan with no custom domain and no credits
> 
> Let me know if that would be helpful to you, and if I'm missing anything about the scope of the PR.

👋 @thehenrybyrd Thanks for looking into this.  What you've outlined is about right.
- domains list is accurate to the extent whatever the API returns
- dashboard UI is fully implemented, except for some updates in [i5](pbArwn-2YJ-p2)

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/domains/DomainsDashboardViewModel.kt:
```diff
@@ -170,21 +164,21 @@ class DomainsDashboardViewModel @Inject constructor(
     private fun onGetDomainClick() {
-        analyticsTrackerWrapper.track(DOMAIN_CREDIT_REDEMPTION_TAPPED, selectedSite)
-        _onNavigation.value = Event(GetDomain(selectedSite))
+        analyticsTrackerWrapper.track(DOMAIN_CREDIT_REDEMPTION_TAPPED, site)
```
This is probably my fault, but this tracking should inside of `onClaimDomainClick` instead (and the `// TODO` comment there should be here).

ravishanker replied:
Fixed, thanks

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/domains/DomainsDashboardViewModel.kt:
```diff
@@ -170,21 +164,21 @@ class DomainsDashboardViewModel @Inject constructor(
     }
 
     private fun onAddDomainClick() {
         onClaimDomainClick()
```
We should check if the user has credits before calling this here, otherwise, the suggestions screen will show up as if the user has credits, even if they are on a free plan.

Edit: to clarify, I'm referring to calling `onClaimDomainClick` inside `onAddDomainClick`.

ravishanker replied:
Updated to check domain credit before invoking

renanferrari commented on WordPress/src/test/java/org/wordpress/android/ui/domains/DomainsDashboardViewModelTest.kt:
```diff
@@ -0,0 +1,279 @@
+import org.wordpress.android.ui.utils.UiString.UiStringRes
+import org.wordpress.android.util.analytics.AnalyticsTrackerWrapper
+
+class DomainsDashboardViewModelTest : BaseUnitTest() {
```
Thank you for adding all these tests! 🙇‍♂️

renanferrari reviewed (commented):
Thank you for the changes! I made sure to test this according to the scenarios we mapped before, and this was the result:

|Scenario|Expected|Actual|
|-----|-----|-----|
|1 ✅|<img width="300" alt="1-expected" src="https://user-images.githubusercontent.com/830056/137177207-445a925f-76e0-4bf3-b4c1-c5ad01721739.png">|<img width="300" alt="1-actual" src="https://user-images.githubusercontent.com/830056/137177179-c7655b70-db98-4e83-a63b-ee855fb94d08.jpeg">|
|2 ✅|<img width="300" alt="2-expected" src="https://user-images.githubusercontent.com/830056/137177208-ccc0d90a-d5ad-4e3d-b112-2bb524380959.png">|<img width="300" alt="2-actual" src="https://user-images.githubusercontent.com/830056/137177187-023e7fff-4868-41c5-a783-9226440e00d2.jpeg">|
|3 ❌|<img width="300" alt="3-expected" src="https://user-images.githubusercontent.com/830056/137177191-d74bbb90-e207-4b5f-9fdd-d070a6b1bb18.png">|<img width="300" alt="3-actual" src="https://user-images.githubusercontent.com/830056/137177190-fb2dc8b9-36cb-4888-8f32-4d6f4405454c.jpeg">|
|4 ❌|<img width="300" alt="4-expected" src="https://user-images.githubusercontent.com/830056/137177209-2c148035-745e-4feb-a42e-2d1dd79b4400.png">|<img width="300" alt="4-actual" src="https://user-images.githubusercontent.com/830056/137177193-764c71ca-bf57-4532-8318-5148b9070968.jpeg">|
|5 ❌|<img width="300" alt="5-expected" src="https://user-images.githubusercontent.com/830056/137177197-cec42879-0681-409e-81dd-1eac0acb2da0.png">|<img width="300" alt="5-actual" src="https://user-images.githubusercontent.com/830056/137177196-5bf58f7f-e16c-4791-bc57-7f2594cd0f6f.jpeg">|
|6 ✅|<img width="300" alt="6-expected" src="https://user-images.githubusercontent.com/830056/137177201-258279df-d177-4726-aeca-dc983113c24b.png">|<img width="300" alt="6-actual" src="https://user-images.githubusercontent.com/830056/137177198-834f5425-9962-4c64-b5d2-b5d5a575e17f.jpeg">|
|7 ❌|<img width="300" alt="7-expected" src="https://user-images.githubusercontent.com/830056/137177210-84c68bf4-241e-4824-ae41-96c9ea48ea0a.png">|<img width="300" alt="7-actual" src="https://user-images.githubusercontent.com/830056/137177203-45977c32-0f2c-4683-8981-faa1403e88fe.jpeg">|
|8 ✅|<img width="300" alt="8-expected" src="https://user-images.githubusercontent.com/830056/137177206-786fd0d5-4a2a-4881-9033-1bbda0523dbe.png">|<img width="300" alt="8-actual" src="https://user-images.githubusercontent.com/830056/137177205-4feabafc-c2fe-4600-a071-390308bfca51.jpeg">|

Aside from some styling differences (spacing, text size, colors, etc) and some new requirements that were missing before (scenarios 3 and 8), which I believe will be handled in a separate PR, I'd say the main issue with the current solution is that the logic to determine whether or not a site has credits is not working. I think that's due to how we're using `DomainRegistrationHandler`, but I'd be ok handling that myself in another PR, as I have more familiarity with that code. Let me know if that's ok with you @ravishanker.

That said, there are a few issues with the current behavior that I believe should be addressed within this PR:

1. The "Free WordPress.com address" card is supposed to show the original site domain (the one that often ends with `wordpress.com`). Instead, it always shows the primary domain. I think there was some confusion there, since even the respective `DomainsDashboardItem` is called `PrimaryDomain` (it should be something like `FreeDomain`).
2. Also related to the issue above, the "Primary site address" label is supposed to be shown next to the primary domain. Instead, it is always shown inside the "Free WordPress.com address" card.
3. Sometimes the custom domain expiration date is set to `false`. I'm not sure if the API is already returning it like this, but if so, then we should at least not propagate it to the UI.

@ravishanker Let me know if this makes sense to you.

@renanferrari  Thank you very much for a detailed comparison 🙇🏻.  Updated to cover the scenarios better.

I also left a few comments below with some other minor suggestions.

ravishanker committed:
fix primary site address label

show primary site address label only when domain is primary

ravishanker committed:
fix site domains list

- show always original site domain at the top
- check domain credit in add domain
- fix tracking

ravishanker committed:
fix domain credit

- removed domain registration handler and call fetch plans
- fix test

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/domains/DomainsDashboardViewModel.kt:
```diff
@@ -89,31 +96,49 @@ class DomainsDashboardViewModel @Inject constructor(
 
-        listItems += getPrimaryDomainItems(siteUrl)
-
+        val freeDomain = domains?.firstOrNull { it.wpcomDomain || it.isWpcomStagingDomain }
```
I wasn't sure this was going to account for free `.blog` domains, but apparently, it does 🙌

<img width=300 src="https://user-images.githubusercontent.com/830056/137521565-41c68298-ca7d-49a3-b5ac-562edd0ea5a9.png"/>

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/domains/DomainsDashboardViewModel.kt:
```diff
@@ -89,31 +96,49 @@ class DomainsDashboardViewModel @Inject constructor(
 
-    private fun getPrimaryDomainItems(siteUrl: String) =
-        listOf(PrimaryDomain(UiStringText(siteUrl), this::onChangeSiteClick))
+    private fun getPrimaryDomainItems(siteUrl: String, isPrimary: Boolean) =
+        listOf(PrimaryDomain(UiStringText(siteUrl), isPrimary, this::onChangeSiteClick))
```
We should probably rename these later to replace the incorrect usage of "primary" with "free".

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/domains/DomainsDashboardViewModel.kt:
```diff
@@ -134,47 +159,60 @@ class DomainsDashboardViewModel @Inject constructor(
         return listItems
     }
 
+    private fun getHomeUrlOrHostName(unmappedUrl: String): String {
+        var homeURL = UrlUtils.removeScheme(unmappedUrl)
+        homeURL = StringUtils.removeTrailingSlash(homeURL)
+        return homeURL
+    }
```
Is there any particular reason why the method on `SiteUtils` was replaced with this? Were there any problems with the logic there?

ravishanker replied:
SiteUtils is using site.url which is returning mapped url not the free url that we need here.  This one is using unmappedUrl which always seem to return the free one.

renanferrari reviewed (approved):
@ravishanker Thanks for the changes! All the main issues I mentioned before seemed to have been fixed and this is looking much better now 🙌 Good job!

Even though I'm still seeing some minor issues (the "Expires on false" issue remains and the "Get your domain" card sometimes appears when it shouldn't), it's nothing that we can't address in a separate PR, so I'm going to go ahead and merge this.

Also left a few comments and questions below for future reference.

LGTM 👍
